parl/remote/remote_decorator.py

```diff
@@ -179,8 +179,77 @@ def request_cpu_resource(self, global_client, max_memory):
                     cnt -= 1
                 return None
 
-            def __getattr__(self, attr):
+            def check_attribute(self, attr):
+                '''checkout if attr is a function or not'''
+                self.internal_lock.acquire()
+                self.job_socket.send_multipart(
+                    [remote_constants.CHECK_ATTRIBUTE,
+                     to_byte(attr)])
+                message = self.job_socket.recv_multipart()
+                self.internal_lock.release()
+                tag = message[0]
+                if tag == remote_constants.NORMAL_TAG:
+                    return loads_return(message[1])
+                else:
+                    self.job_shutdown = True
+                    raise NotImplementedError()
+
+            def set_remote_attr(self, attr, value):
+                self.internal_lock.acquire()
+                self.job_socket.send_multipart([
+                    remote_constants.SET_ATTRIBUTE,
+                    to_byte(attr),
+                    dumps_return(value)
+                ])
+                message = self.job_socket.recv_multipart()
+                tag = message[0]
+                self.internal_lock.release()
+                if tag == remote_constants.NORMAL_TAG:
+                    pass
+                else:
+                    self.job_shutdown = True
+                    raise NotImplementedError()
+                return
+
+            def get_remote_attr(self, attr):
                 """Call the function of the unwrapped class."""
+                #check if attr is a function or not
+                is_attribute = self.check_attribute(attr)
+                if is_attribute:
+                    self.internal_lock.acquire()
+                    self.job_socket.send_multipart(
+                        [remote_constants.GET_ATTRIBUTE,
+                         to_byte(attr)])
+                    message = self.job_socket.recv_multipart()
+                    tag = message[0]
+
+                    if tag == remote_constants.NORMAL_TAG:
```

Comment:
line 224-252 is repeated to line 266-296, can we provide a common function.

Reply:
modified